### PR TITLE
Feature/fga/simplify timeline logic

### LINKED ISSUE #5166: Repeated timeline, many read receipts
### Steps to reproduce

Log in to a conduit account.
Accept a federated invite into a new room.

The problem seems to be related to the limited flag for syncs or the /messages endpoint.

### Your phone model

-

### Operating system version

-

### Application version and app store

-

### Homeserver

conduit.rs

### Will you send logs?

No

### LINKED ISSUE #6312: Joining a DM shows infinite room join events in the timeline until app crashes
### Steps to reproduce

I don't know how to trigger this bug, but it was reproducible until I opened the app

1. Other person sends some messages in the DM
2. Receive an invite to a DM and click the notification
3. See the invite dialog open in the app (it doesn't say who the invite is from, the title of the dialog is `!room1d:homeserver.org`
4. Click "Join"
5. App crashes

I could reproduce the behaviour until I opened the DM in the web app, then it started working correctly. I tried submitting a rageshake after the app restarted and it claimed to have submitted correctly but I can't see it in the rageshake repo.

#### What did you expect?
See DM

#### What happened instead?
Room opens, see a "membership change" event in the timeline, see number of the membership change events counting up. App crashes.

https://user-images.githubusercontent.com/51663/173808555-bcc7d803-75a9-4028-b7ff-600ede9f6900.mp4

### Your phone model

Pixel 2

### Operating system version

Android 11

### Application version and app store

Element version 1.4.20; Matrix SDK 1.4.20, olm version 3.2.11

### Homeserver

matrix.org, the other person is on element.io

### Will you send logs?

Yes

### Are you willing to provide a PR?

No

### PR DESCRIPTION
**Pull request by ganfra**
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Simplify the management of timeline events and chunks : 
- Always delete all chunks of the room when having a gappy sync 
- Always insert timeline events in a chunk and don't try to link overlapping chunks

## Motivation and context
There was multiple scenarios where we had missing messages or loops in timeline. 
This fixes #6312 and fixes #5166 

## Tested devices

- [x] Physical
- [x] Emulator
- OS version(s):

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/handler/room/RoomSyncHandler.kt [hunk 1]
```diff
@@ -381,12 +382,13 @@ internal class RoomSyncHandler @Inject constructor(
...
         val lastChunk = ChunkEntity.findLastForwardChunkOfRoom(realm, roomEntity.roomId)
-        if (isLimited && lastChunk != null) {
-            lastChunk.deleteOnCascade(deleteStateEvents = false, canDeleteRoot = true)
-        }
         val chunkEntity = if (!isLimited && lastChunk != null) {
             lastChunk
         } else {
+            // Delete all chunks of the room in case of gap.
+            ChunkEntity.findAll(realm, roomId).forEach {
+                it.deleteOnCascade(false, canDeleteRoot = true)
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/handler/room/RoomSyncHandler.kt:390]: `canDeleteRoot = true` is there a risk to delete state event here? Or Event needed for aggregation?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/migration/MigrateSessionTo030.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    override fun doMigrate(realm: DynamicRealm) {
+        // Delete all previous chunks
+        val chunks = realm.where("ChunkEntity")
+                .equalTo(ChunkEntityFields.IS_LAST_FORWARD, false)
+                .findAll()
+
+        chunks.forEach { chunk ->
+            chunk.getList(ChunkEntityFields.TIMELINE_EVENTS.`$`).clearWith { timelineEvent ->
+                // Don't delete state events
+                if (timelineEvent.isNull(TimelineEventEntityFields.ROOT.TYPE)) {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/migration/MigrateSessionTo030.kt:40]: I think you mean

```suggestion
                if (timelineEvent.isNull(TimelineEventEntityFields.ROOT.STATE_KEY)) {
```

**Review by bmarty — CHANGES_REQUESTED** [review]

**Review by bmarty — APPROVED** [review]
LGTM

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/migration/MigrateSessionTo030.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        // Delete all previous chunks
+        val chunks = realm.where("ChunkEntity")
+                .equalTo(ChunkEntityFields.IS_LAST_FORWARD, false)
+                .findAll()
+
+        chunks.forEach { chunk ->
+            chunk.getList(ChunkEntityFields.TIMELINE_EVENTS.`$`).clearWith { timelineEvent ->
+                // Don't delete state events
+                if (timelineEvent.isNull(TimelineEventEntityFields.ROOT.STATE_KEY)) {
+                    timelineEvent.getObject(TimelineEventEntityFields.ROOT.`$`)?.deleteFromRealm()
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/migration/MigrateSessionTo030.kt:41]: Yes we can keep them, they are kind of independent

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TimelineChunk.kt [hunk 4]
```diff
@@ -445,7 +445,7 @@ internal class TimelineChunk(
             Timber.e(failure, "Failed to fetch from server")
             LoadMoreResult.FAILURE
         }
-        return if (loadMoreResult == LoadMoreResult.SUCCESS) {
+        return if (loadMoreResult != LoadMoreResult.FAILURE) {
```
Comment (mnaturel) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TimelineChunk.kt:448]: Just checking, is it intended to call `loadMore` when result is `LoadMoreResult.REACHED_END`?

**Review by mnaturel — APPROVED** [review]
Good for me. I have just added one question about something I was not sure.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TimelineChunk.kt (hunk 4 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TimelineChunk.kt:448]: This is the downside of linking the last permalink chunk to the lastForward chunk... I still don't know if we want this or just add timeline events from sync on both chunks instead

File: changelog.d/6318.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Fix loop in timeline and simplify management of chunks and timeline events.
```
Comment (MadLittleMods) [inline changelog.d/6318.bugfix:1]: Does this also close the following? They seem like duplicates of https://github.com/vector-im/element-android/issues/5166 anway

 - https://github.com/vector-im/element-android/issues/5690
 - https://github.com/vector-im/element-android/issues/6342

**Pull request merged**